### PR TITLE
Initial implementation of SavedModel I/O for Graph Transform Tool

### PR DESCRIPTION
**Pull request by frreiss**
This PR adds a Python utility for reading SavedModels into the Graph Transform Tool, as requested in #20519.

The change set adds a function `tensorflow.tools.graph_transforms.TransformSavedModel`. This function takes as input the location of a SavedModel, some parameters for extracting a single inference function from the SavedModel, and a list of transforms to run. `TransformSavedModel` loads the SavedModel, extracts the graph for the specified function, runs this graph through the Graph Transform Tool, and writes out a new SavedModel containing the output of the tool.

`TransformSavedModel` works on both V1 (graph-based) and V2 (function-based) SavedModel formats. Regardless of the format of the input SavedModel, the output of `TransformSavedModel` is always in the newer V2 format.

Here's some example usage:
```python
import tensorflow as tf
from tensorflow.tools import graph_transforms

tf.enable_eager_execution()

# Define a simple model with an extra op
class MyTrackable(tf.train.Checkpoint):
    @tf.function(input_signature=[tf.TensorSpec(shape=None, dtype=tf.int32),
                                  tf.TensorSpec(shape=None, dtype=tf.int32)])
    def test_func(self, first_arg, second_arg):
        result1 = first_arg + second_arg
        result2 = first_arg - second_arg
        tf.constant(42., name="dead_code")
        return {"sum": result1, "difference": result2}
    
before_dir = "./model_before"
after_dir = "./model_after"

tf.saved_model.save(MyTrackable(), before_dir)

one = tf.constant(1, dtype=tf.int32)
two = tf.constant(2, dtype=tf.int32)
model_before = tf.saved_model.load_v2(before_dir)
print("Result before: {}".format(
    model_before.signatures["serving_default"](first_arg=two, second_arg=one)))
print("Contains dead code before: {}".format( 
    str("dead_code" in str(model_before.signatures["serving_default"].graph.as_graph_def()))))

graph_transforms.TransformSavedModel(before_dir, after_dir, ["strip_unused_nodes(type=float32)"])
model_after = tf.saved_model.load_v2(after_dir)
print("Result after: {}".format(
    model_after.signatures["serving_default"](first_arg=two, second_arg=one)))
print("Contains dead code after: {}".format( 
    str("dead_code" in str(model_after.signatures["serving_default"].graph.as_graph_def()))))
```
The above code produces the following output:
```
Result before: {'sum': <tf.Tensor: id=2042, shape=(), dtype=int32, numpy=3>, 'difference': <tf.Tensor: id=2041, shape=(), dtype=int32, numpy=1>}
Contains dead code before: True
Result after: {'sum': <tf.Tensor: id=2368, shape=(), dtype=int32, numpy=3>, 'difference': <tf.Tensor: id=2367, shape=(), dtype=int32, numpy=1>}
Contains dead code after: False
```

To keep this change set (relatively) small, I have limited the scope of this version of `TransformSavedModel`. Specifically, the current implementation:
* does not support SavedModels that contain variables
* does not support V2 SavedModels in which the inference code is spread across multiple TensorFlow functions
* does not pass through auxiliary fields and alternate entry points of the `Trackable` object in a V2 SavedModel; only the user's selected signature is retained in the output SavedModel file.

This implementation **does** support V2 SavedModels consisting of a single "core" inference function that is called from a chain of one or more "wrapper" functions; where the "wrapper" functions only translate the format of input and output arguments. The example Python code above produces such a SavedModel.

The implementation of `TransformSavedModel` is in the file `tensorflow/tools/graph_transforms/transform_saved_model.py`. I also added automated regression tests to `tensorflow/tools/graph_transforms/python/transform_graph_test.py`. I also moved the existing `TransformGraph` function to its own Python file.

### CODE REVIEW

**Comment by gbaned** [comment]
@petewarden Could you PTAL and approve.

**Review by petewarden — DISMISSED** [review]

**Comment by gbaned** [comment]
@frreiss Could you please check failed build errors? Thanks!

**Comment by frreiss** [comment]
Looking into the test failures in the builds.

**Comment by frreiss** [comment]
Looks like an idiosyncrasy in `function_def_to_graph`'s node naming that I had been working around got fixed recently. I've removed the workaround. Tests are passing again on my test machines.

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by gbaned** [comment]
@petewarden Can you please take a look on this PR? Thanks!

**Comment by gbaned** [comment]
@frreiss Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@petewarden Can you please take a look on this PR? Thanks!

**Comment by HananKavitz** [comment]
@frreiss 
Is this expected to be approved?
Any idea when?

thx

**Comment by frreiss** [comment]
@HananKavitz I am not optimistic at this point. I'll leave it open for a few more weeks though.

**Review by av8ramit — APPROVED** [review]

**Comment by av8ramit** [comment]
So I'm okay with this from the Python side of things. But I'll leave the functionality decision up to @petewarden who seems to own these files.

**Review by av8ramit — DISMISSED** [review]

File: tensorflow/tools/graph_transforms/transform_graph.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# ==============================================================================
+"""Exposes the Python wrapper for graph transforms over GraphDef protocol
+buffer messages."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from tensorflow.core.framework import graph_pb2
+from tensorflow.python.pywrap_tensorflow import TransformGraphWithStringInputs
+from tensorflow.python.util import compat
```
Comment (rthadur) [inline tensorflow/tools/graph_transforms/transform_graph.py]: `third_party/tensorflow/tools/graph_transforms/transform_graph.py", line 24, in <module>
    from google3.third_party.tensorflow.python.pywrap_tensorflow import TransformGraphWithStringInputs
ImportError: cannot import name 'TransformGraphWithStringInputs'`

@frreiss can you please fix this error

**Review by rthadur — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@frreiss Can you please check @rthadur's comments and keep us posted. Thanks!

File: tensorflow/tools/graph_transforms/transform_graph.py (hunk 1 above)
Comment (frreiss) [reply tensorflow/tools/graph_transforms/transform_graph.py:23]: This should be fixed now.

**Comment by gbaned** [comment]
Hi @av8ramit, Can you please review this PR ? Thanks!

**Comment by av8ramit** [comment]
@gbaned unfortunately, I'm not the best person to take a look at this. Perhaps @petewarden or @bmzhao?

**Comment by gbaned** [comment]
@petewarden, @bmzhao Can you please review this PR ? Thanks!

**Review by petewarden — APPROVED** [review]

File: tensorflow/tools/graph_transforms/transform_saved_model.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+from tensorflow.core.framework import graph_pb2
+from tensorflow.python.eager import context
+from tensorflow.python.eager import function
+from tensorflow.python.framework import func_graph
+from tensorflow.python.framework import function_def_to_graph
+from tensorflow.python.framework import graph_to_function_def
+from tensorflow.python.framework import importer
+from tensorflow.python.framework import ops
+from tensorflow.python.saved_model import load
```
Comment (rthadur) [inline tensorflow/tools/graph_transforms/transform_saved_model.py:28]: @frreiss can you please check below error 

`third_party/tensorflow/tools/graph_transforms/transform_saved_model.py", line 28, in <module>
    from google3.third_party.tensorflow.python.saved_model import load
ImportError: cannot import name 'load'
-- Forge runner: Test failed`

**Review by rthadur — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@frreiss Can you please check @rthadur's comments and keep us posted. Thanks!

**Comment by frreiss** [comment]
Looking into the test failure.

**Comment by gbaned** [comment]
@frreiss Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
@frreiss Any update on this PR? Please. Thanks!

**Comment by frreiss** [comment]
I'm unable to reproduce the problem that is occurring in the internal Google build.

When I run:

```
bazel test --config=v1 //tensorflow/tools/graph_transforms:transform_graph_py_test
```

everything looks ok:

```
Target //tensorflow/tools/graph_transforms:transform_graph_py_test up-to-date:
  bazel-bin/tensorflow/tools/graph_transforms/transform_graph_py_test
INFO: Elapsed time: 6379.650s, Critical Path: 653.87s
INFO: 6792 processes: 6792 local.
INFO: Build completed successfully, 7034 total actions
//tensorflow/tools/graph_transforms:transform_graph_py_test              PASSED in 28.8s

INFO: Build completed successfully, 7034 total actions
```

If I remove the "v1only" tag from the `transform_graph_py_test` target, I can run:

```
bazel test //tensorflow/tools/graph_transforms:transform_graph_py_test
```

(without the `--config=v1`) and everything goes ok.

I tried building a wheel with the "v1" APIs for manual testing, but that part of the build seems to be broken right now. 

Overall, I've been maintaining the branch for this PR has for 15 months now.  The maintenance has taken a considerable amount of time. I've had to do multiple manual merges, and each build/test cycle takes an entire day. I am closing this PR. I'll leave the branch on my fork for a few more months in case anyone needs the code for pulling the graph out of a V2 SavedModel.

**Pull request closed without merging**